### PR TITLE
Sc 31 testcases for evm atmoic tx for defi cli

### PR DESCRIPTION
**Pull request by akp111**

### CODE REVIEW

File: src/cli.ts [hunk 1]
```diff
@@ -297,8 +317,11 @@ export async function initCtxJsonFromOptions(options: OptionValues): Promise<voi
 
 //////////////////////////////////////////////////////////////////////////////////////////
 // Network info
-
-function logAddressInfo(ctx: Context) {
+/** @description - Logs address info
+ * @description Returns the address info
```
Comment (avisingh2) [inline src/cli.ts]: replace @description with @returns

File: src/cli.ts [hunk 2]
```diff
@@ -148,7 +148,12 @@ export async function cli(program: Command) {
     })
 }
 
-async function contextFromOptions(options: OptionValues): Promise<Context> {
+/**
+ * Returns the context based the source passed in the options
```
Comment (avisingh2) [inline src/cli.ts]: move this statement against @returns

File: src/evmAtomicTx.ts [hunk 3]
```diff
@@ -107,7 +107,7 @@ export async function issueSignedEvmTxPCImport(ctx: Context, signedTxJson: Signe
     ctx.cchain.buildImportTx(...deserializeImportPC_args(serialization)))
 }
 
-async function issueSignedEvmTx(ctx: Context, signedTxJson: SignedTxJson,
+export async function issueSignedEvmTx(ctx: Context, signedTxJson: SignedTxJson,
```
Comment (avisingh2) [inline src/evmAtomicTx.ts:117]: please add documentation for all the exported function

File: src/evmAtomicTx.ts [hunk 4]
```diff
@@ -117,7 +117,7 @@ async function issueSignedEvmTx(ctx: Context, signedTxJson: SignedTxJson,
   return { chainTxId: chainTxId }
 }
 
-async function getExportCPParams(ctx: Context, amount: BN, fee?: BN, nonce?: number, threshold: number = 1): Promise<ExportCPParams> {
+export async function getExportCPParams(ctx: Context, amount: BN, fee?: BN, nonce?: number, threshold: number = 1): Promise<ExportCPParams> {
```
Comment (avisingh2) [inline src/evmAtomicTx.ts:136]: please add documentation for all the exported function

File: src/evmAtomicTx.ts [hunk 5]
```diff
@@ -135,15 +135,15 @@ async function getExportCPParams(ctx: Context, amount: BN, fee?: BN, nonce?: num
     threshold,
     fee ?? baseFee
   ]
-  const unsignedTx: UnsignedTx = await ctx.cchain.buildExportTx(...params)
-  if (fee === undefined) {
+   if (fee === undefined) {
```
Comment (avisingh2) [inline src/evmAtomicTx.ts]: instead of comparing with undefined why not check with `if(!fee)` like this

File: src/evmAtomicTx.ts [hunk 6]
```diff
@@ -135,15 +135,15 @@ async function getExportCPParams(ctx: Context, amount: BN, fee?: BN, nonce?: num
...
+    const unsignedTx: UnsignedTx = await ctx.cchain.buildExportTx(...params)
     const exportCost: number = costExportTx(unsignedTx)
     params[9] = baseFee.mul(new BN(exportCost))
   }
   return params
 }
 
-async function getImportPCParams(ctx: Context, fee?: BN): Promise<ImportPCParams> {
+export async function getImportPCParams(ctx: Context, fee?: BN): Promise<ImportPCParams> {
```
Comment (avisingh2) [inline src/evmAtomicTx.ts:168]: please add documentation for all the exported function

File: src/evmAtomicTx.ts [hunk 7]
```diff
@@ -159,8 +159,8 @@ async function getImportPCParams(ctx: Context, fee?: BN): Promise<ImportPCParams
     [ctx.cAddressBech32!],
     baseFee
   ]
-  const unsignedTx: UnsignedTx = await ctx.cchain.buildImportTx(...params)
   if (fee === undefined) {
```
Comment (avisingh2) [inline src/evmAtomicTx.ts]: change it to `if(!fee)`

File: test/fixtures/addDelegate.data.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import { BN } from '@flarenetwork/flarejs/dist';
+const fixtures = {
```
Comment (avisingh2) [inline test/fixtures/addDelegate.data.ts:3]: put blank line when all imports end

File: test/fixtures/addValidator.data.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import { BN } from '@flarenetwork/flarejs/dist';
+const fixtures = {
```
Comment (avisingh2) [inline test/fixtures/addValidator.data.ts:3]: add blank line

File: test/unit/forDefi.test.ts [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+            id: fixtures.sendToForDefi.mock.txId
+          }),
+          {
+            status: 200,
+            statusText: 'OK'
+          }
+        )
+      );
+
+      // const result = await forDefi.sendToForDefi(
```
Comment (avisingh2) [inline test/unit/forDefi.test.ts]: remove the commented code if not needed

File: test/unit/pvmAtomicTx.test.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  exportTxPC,
+  importTxCP,
+  issueSignedPvmTx
+} from '../../src/pvmAtomicTx';
+import { contextEnv } from '../../src/constants';
+import { Context } from '../../src/interfaces';
+import fixtures from '../fixtures/pvmAtomicTx.data';
+import { UTXOSet, UnsignedTx, Tx } from '@flarenetwork/flarejs/dist/apis/platformvm';
+import { EcdsaSignature } from '@flarenetwork/flarejs/dist/common';
+describe('pvmAtomicTx testcases', () => {
```
Comment (avisingh2) [inline test/unit/pvmAtomicTx.test.ts:15]: add blank line

File: test/unit/utils.test.ts [hunk 12]
```diff
@@ -36,7 +36,7 @@ import {
 } from '../../src/utils';
 import { UnsignedTx as EvmUnsignedTx, UTXOSet } from '@flarenetwork/flarejs/dist/apis/evm';
 import { UnsignedTx as PvmUnsignedTx } from '@flarenetwork/flarejs/dist/apis/platformvm';
-import fixtures from '../fixtures/utilsData';
+import fixtures from '../fixtures/utils.data';
 import { serialize, covertBNToSting, compareValues } from '../helper/testHelpers';
 describe('Unit Test Cases for utils', () => {
```
Comment (avisingh2) [inline test/unit/utils.test.ts:42]: add blank line

File: src/addDelegator.ts [hunk 13]
```diff
@@ -59,7 +59,16 @@ export async function getUnsignedAddDelegator(
   }
 }
 
-async function getAddDelegatorParams(
+
+/**
+ * Generate the parameters for getUnsignedAddDelegator and addDelegator
```
Comment (avisingh2) [inline src/addDelegator.ts]: add the @description

File: src/addValidator.ts [hunk 14]
```diff
@@ -63,7 +63,17 @@ export async function getUnsignedAddValidator(
   }
 }
 
-async function getAddValidatorParams(
+
+/**
+ * Get the parameters for getUnsignedAddValidator and addValidator
```
Comment (avisingh2) [inline src/addValidator.ts]: @description

File: src/cli.ts [hunk 15]
```diff
@@ -188,8 +203,13 @@ function getOptions(program: Command, options: OptionValues): OptionValues {
   }
   return { ...allOptions, network }
 }
-
-function capFeeAt(cap: number, usedFee?: string, specifiedFee?: string): void {
+/**
+ *
```
Comment (avisingh2) [inline src/cli.ts]: again add @description

File: src/cli.ts [hunk 16]
```diff
@@ -307,7 +331,11 @@ function logAddressInfo(ctx: Context) {
   log(`secp256k1 public key: 0x${compressedPubKey}`)
 }
 
-async function logBalanceInfo(ctx: Context) {
+/**
+ * @description Returns the balance info of the account
```
Comment (avisingh2) [inline src/cli.ts]: you should put function description in @description and returns summary in @returns

File: src/cli.ts [hunk 17]
```diff
@@ -326,7 +358,11 @@ function logNetworkInfo(ctx: Context) {
   log(`assetId: ${ctx.avaxAssetID}`)
 }
 
-async function logValidatorInfo(ctx: Context) {
+/**
+ * @description Returns the validator information regrading current and pending validators
```
Comment (avisingh2) [inline src/cli.ts]: same here returns summary will be against @returns

File: src/constants.ts [hunk 18]
```diff
@@ -12,11 +12,22 @@ import {
   decodePublicKey
 } from './utils'
 
-function readContextFile(ctxFile: string): ContextFile {
+/**
+ * @description - parses the file and returns the context of ctx.json
+ * @param ctxFile - path to the context file
+ * @returns
```
Comment (avisingh2) [inline src/constants.ts]: @returns what?

File: src/constants.ts [hunk 19]
```diff
@@ -26,17 +37,27 @@ export function contextEnv(path: string, network: string): Context {
     process.env.PRIVATE_KEY_CB58)
 }
 
+/**
+ *
```
Comment (avisingh2) [inline src/constants.ts]: description?

File: src/constants.ts [hunk 20]
```diff
@@ -26,17 +37,27 @@ export function contextEnv(path: string, network: string): Context {
...
+ * @param ctxFile - path to the context file
+ * @returns returns the context
+ */
 export function contextFile(ctxFile: string): Context {
   const ctx = readContextFile(ctxFile)
   return getContext(ctx.network, ctx.publicKey)
 }
 
+/**
+ *
```
Comment (avisingh2) [inline src/constants.ts]: @description ?

File: src/constants.ts [hunk 21]
```diff
@@ -46,20 +67,35 @@ export const colorCodes = {
...
   "Coston2": "costwo"
 }
 
+/**
+ * @description returns the context
+ * @param network - network name: flare/localflare/costwo
+ * @param publicKey - public key
+ * @param privateKeyHex - private key in hex format
+ * @param privateKeyCB58 - private key in cb58 format
+ * @returns
```
Comment (avisingh2) [inline src/constants.ts]: if returns void than remove the @returns annotation

File: src/evmAtomicTx.ts [hunk 22]
```diff
@@ -107,7 +107,14 @@ export async function issueSignedEvmTxPCImport(ctx: Context, signedTxJson: Signe
     ctx.cchain.buildImportTx(...deserializeImportPC_args(serialization)))
 }
 
-async function issueSignedEvmTx(ctx: Context, signedTxJson: SignedTxJson,
+/**
+ *
```
Comment (avisingh2) [inline src/evmAtomicTx.ts]: @description ?

File: src/evmAtomicTx.ts [hunk 23]
```diff
@@ -117,7 +124,16 @@ async function issueSignedEvmTx(ctx: Context, signedTxJson: SignedTxJson,
   return { chainTxId: chainTxId }
 }
 
-async function getExportCPParams(ctx: Context, amount: BN, fee?: BN, nonce?: number, threshold: number = 1): Promise<ExportCPParams> {
+/**
+ *
```
Comment (avisingh2) [inline src/evmAtomicTx.ts]: @description?

File: src/evmAtomicTx.ts [hunk 24]
```diff
@@ -135,15 +151,21 @@ async function getExportCPParams(ctx: Context, amount: BN, fee?: BN, nonce?: num
...
+    const unsignedTx: UnsignedTx = await ctx.cchain.buildExportTx(...params)
     const exportCost: number = costExportTx(unsignedTx)
     params[9] = baseFee.mul(new BN(exportCost))
   }
   return params
 }
 
-async function getImportPCParams(ctx: Context, fee?: BN): Promise<ImportPCParams> {
+/**
+ *
```
Comment (avisingh2) [inline src/evmAtomicTx.ts]: @description?

File: src/forDefi.ts [hunk 25]
```diff
@@ -7,17 +6,24 @@ import { SignedTxJson, SignedWithdrawalTxJson, UnsignedTxJson, UnsignedWithdrawa
 
 const gatewayHost = "api.fordefi.com"
 
-export async function sendToForDefi(unsignedTxidFile: string, ctxFile: string, withdrawal: boolean = false): Promise<string> {
+/**
+ * Send signature to forDefi
```
Comment (avisingh2) [inline src/forDefi.ts]: @description?

File: src/forDefi.ts [hunk 26]
```diff
@@ -7,17 +6,24 @@ import { SignedTxJson, SignedWithdrawalTxJson, UnsignedTxJson, UnsignedWithdrawa
...
+ * @param unsignedTxidFile - path to the file
+ * @param ctxFile - ctx file
+ * @param withdrawal - boolen if its a withdrawl trx or not
+ * @param _getVaultPublickey - for testcase mocking purpose, bydefault it calls getVaultPublickey
+ * @returns
```
Comment (avisingh2) [inline src/forDefi.ts:15]: returns what?

File: src/forDefi.ts [hunk 27]
```diff
@@ -116,12 +122,16 @@ export async function getSignature(unsignedTxidFile: string, withdrawal: boolean
     return signatureHex;
 }
 
-async function getVaultPublickey(vaultId: string): Promise<string> {
+/**
```
Comment (avisingh2) [inline src/forDefi.ts:131]: @description?

File: src/pvmAtomicTx.ts [hunk 28]
```diff
@@ -83,7 +83,13 @@ export async function issueSignedPvmTx(ctx: Context, signedTxJson: SignedTxJson)
   return { chainTxId: chainTxId }
 }
 
-async function getImportCPParams(ctx: Context, threshold: number = 1): Promise<ImportCPParams> {
+/**
+ * Returns the parameters related to importing funds from C to P chain
```
Comment (avisingh2) [inline src/pvmAtomicTx.ts]: @description?

File: test/helper/testHelpers.ts [hunk 29]
```diff
@@ -10,14 +18,30 @@ export const serialize = (bufferData: Buffer[]) => {
   return serializedData;
 };
 
+/**
+ *
```
Comment (avisingh2) [inline test/helper/testHelpers.ts]: @description?

File: test/helper/testHelpers.ts [hunk 30]
```diff
@@ -10,14 +18,30 @@ export const serialize = (bufferData: Buffer[]) => {
...
+ * @param data - data of type any
+ * @returns - returns it in big number format
+ */
 export const covertBNToSting = (data: any): string => {
   return data instanceof BN ? data.toString() : data;
 };
 
+/**
+ *
```
Comment (avisingh2) [inline test/helper/testHelpers.ts]: @description?

File: test/helper/testHelpers.ts [hunk 31]
```diff
@@ -10,14 +18,30 @@ export const serialize = (bufferData: Buffer[]) => {
...
+ * @param arr - array of entities
+ * @returns - converts each to big number and returns the array
+ */
 export const convertArrayBNToString = (arr: any[]): any[] => {
   return arr.map((item) => (item instanceof BN ? item.toString() : item));
 };
 
+/**
+ *
```
Comment (avisingh2) [inline test/helper/testHelpers.ts]: @description?

File: test/helper/testHelpers.ts [hunk 32]
```diff
@@ -33,3 +57,15 @@ export const compareValues = (a: any, b: any): boolean => {
   }
   return a === b;
 };
+
+/**
+ * @description - trnafers funds from C to P chain
```
Comment (avisingh2) [inline test/helper/testHelpers.ts]: typo trnafers please fix it

File: test/unit/evmAtomicTx.test.ts [hunk 33]
```diff
@@ -0,0 +1,230 @@
...
+  issueSignedEvmTx,
+  getUnsignedImportTxPC,
+  getUnsignedExportTxCP,
+  importTxPC,
+  exportTxCP
+} from '../../src/evmAtomicTx';
+import { contextEnv } from '../../src/constants';
+import { Context } from '../../src/interfaces';
+import fixtures from '../fixtures/evmAtomicTx.data';
+describe('evmAtomicTx Testcases', () => {
```
Comment (avisingh2) [inline test/unit/evmAtomicTx.test.ts:14]: add blank line

File: src/cli.ts (hunk 16 above)
Comment (akp111) [reply src/cli.ts]: It does not return anything

File: test/helper/testHelpers.ts [hunk 34]
```diff
@@ -1,4 +1,12 @@
 import { BN } from '@flarenetwork/flarejs/dist';
+import { exportTxCP } from '../../src/evmAtomicTx';
+import { contextEnv } from '../../src/constants';
+
+/**
+ *
```
Comment (avisingh2) [inline test/helper/testHelpers.ts]: description?

**Review by avisingh2 — APPROVED** [review]

**Pull request merged**